src/query/v2/shard_request_manager.hpp

```diff
@@ -0,0 +1,361 @@
+// Copyright 2022 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <chrono>
+#include <deque>
+#include <iostream>
+#include <map>
+#include <optional>
+#include <random>
+#include <set>
+#include <stdexcept>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+
+#include "coordinator/coordinator_client.hpp"
+#include "coordinator/coordinator_rsm.hpp"
+#include "coordinator/shard_map.hpp"
+#include "io/address.hpp"
+#include "io/errors.hpp"
+#include "io/rsm/raft.hpp"
+#include "io/rsm/rsm_client.hpp"
+#include "io/rsm/shard_rsm.hpp"
+#include "io/simulator/simulator.hpp"
+#include "io/simulator/simulator_transport.hpp"
+#include "query/v2/requests.hpp"
+#include "storage/v3/id_types.hpp"
+#include "utils/result.hpp"
+
+namespace requests {
+template <typename TStorageClient>
+class RsmStorageClientManager {
+ public:
+  using CompoundKey = memgraph::io::rsm::ShardRsmKey;
+  using Shard = memgraph::coordinator::Shard;
+  using LabelId = memgraph::storage::v3::LabelId;
+  RsmStorageClientManager() = default;
+  RsmStorageClientManager(const RsmStorageClientManager &) = delete;
+  RsmStorageClientManager(RsmStorageClientManager &&) = delete;
+
+  void AddClient(const LabelId label_id, Shard key, TStorageClient client) {
+    cli_cache_[label_id].insert({std::move(key), std::move(client)});
+  }
+
+  bool Exists(const LabelId label_id, const Shard &key) { return cli_cache_[label_id].contains(key); }
+
+  void PurgeCache() { cli_cache_.clear(); }
+
+  TStorageClient &GetClient(const LabelId label_id, const Shard &key) { return cli_cache_[label_id].find(key)->second; }
+
+ private:
+  std::map<LabelId, std::map<Shard, TStorageClient>> cli_cache_;
+};
+
+template <typename TRequest>
+struct ExecutionState {
```

Comment:
Add in the comment what's the point with the state? Is `ExecutionState` part of `ExecutionContext` while `ShardRequestManager` is part of `InterpreterContext`?

Reply:
The state object should be created per cursor, e.g. `ScanAll` will have the state